### PR TITLE
udev: fixes for device-mapper rules

### PR DESCRIPTION
**Pull request by mwilck**
This PR contains several minor but important fixes to systemd's udev rules for dm devices.

* `ENV{DM_SUSPENDED}` is a device-mapper internal flag that shouldn't be used by non-dm rules. It will be removed in the forcecoming dm ruleset with `ENV{DM_UDEV_RULES_VSN}=="3" (https://gitlab.com/lvmteam/lvm2/-/merge_requests/9).
* The "output" flag of dm that tells later rules to ignore a given device is `ENV{DM_UDEV_DISABLE_OTHER_RULES_FLAG}`. This flag means that no probing attempts should be made on the device, and that relevant udev properties should be imported from the db. It does *not* imply `ENV{SYSTEMD_READY}="0"`, which might lead to the device being deactivated in systemd, and thus unmounted or otherwise deconstructed.
* `DM_UDEV_DISABLE_OTHER_RULES_FLAG` will be 1 if a dm device is suspended, and in  other cases in which the device is not ready for being probed.
* The current code evaluated `DM_UDEV_DISABLE_OTHER_RULES_FLAG` only for "add" events. For dm, "change" events are more important than "add" events.
* It isn't necessary, and actually possibly wrong, to blindly set `ENV{SYSTEMD_READY}="1"`.
* The current rule set jumped to `systemd_end` for suspended dm devices, which is unnecessary, too, because no other rule applies to device-mapper block devices.

Note that while this PR is related to [lvm2#9](https://gitlab.com/lvmteam/lvm2/-/merge_requests/9), it doesn't depend on it, nor vice-versa.

For information: @prajnoha, @fbui, @bmarzins, @msekletar, @jiayi0118, @lnykryn, @yuwata

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
@mwilck It'd be great if https://gitlab.com/lvmteam/lvm2/-/merge_requests/5 could get some attention.

**Comment by mwilck** [comment]
> @mwilck It'd be great if https://gitlab.com/lvmteam/lvm2/-/merge_requests/5 could get some attention.

I'm not one of the LVM2 maintainers. I've made a positive comment. Hope that helps. Perhaps you should propose this patch on the linux-lvm mailing list first; that's what I did.

File: rules.d/99-systemd.rules.in
Comment (yuwata) [inline rules.d/99-systemd.rules.in:32]: Then, when the device get `SYSTEMD_READY=1`?

**Review by yuwata — COMMENTED** [review]
The first two commits looks reasonable. Not sure about the third commit.

File: rules.d/99-systemd.rules.in
Comment (arvidjaar) [reply rules.d/99-systemd.rules.in:32]: > Then, when the device get `SYSTEMD_READY=1`?

Devices are ready by default, so `ENV{SYSTEMD_READY}=1` is really needed only to cancel previous `ENV{SYSTEMD_READY}==0`.

Comment (yuwata) [reply rules.d/99-systemd.rules.in:32]: Ah, yes. Indeed. You are right.

Comment (yuwata) [reply rules.d/99-systemd.rules.in:32]: Then, so the third commit is mostly do nothing, right? If so, the third commit looks also good to me.

**Comment by prajnoha** [comment]
Hmm, but we should still avoid marking the DM device as `SYSTEMD_READY=1` in case we have any uevent before DM dev activation (so before it has any table loaded and resumed, e.g. the genuine `add` event for the DM device). This is not covered by current changes, is it?

File: rules.d/99-systemd.rules.in
Comment (mwilck) [reply rules.d/99-systemd.rules.in:32]: My commit message is misleading. systemd itself reads `SYSTEMD_READY` as a boolean. All values except those that it doesn't parse as "false" (i.e. `0`, `no`, `n`, `false`, `f`, or `off`) cause systemd to assume the device is "ready". This includes the case where `SYSTEMD_READY` is not set.

So indeed, as far as systemd is concerned, `ENV{SYSTEMD_READY}=="1"` is equivalent with the variable being unset, and my assertion that setting this variable to 1 may be harmful is false. I'll fix that.

For udev itself, there's a subtle difference, which @jiayi0118 pointed out in c1a2ada89708d6aeeada496712cb24a4a58e75cc: `IMPORT{db}` is a matching token which evaluates to `true` if and only if the environment variable is set in the db. Setting `ENV{SYSTEMD_READY}="1"` makes sure that future compound rules containing `IMPORT{db}="SYSTEMD_READY"` will be applied.

IMO this is a higly surprising feature of the udev syntax, and I'd generally recommend to avoid using `IMPORT{db}` in compound rules.

**Comment by mwilck** [comment]
@prajnoha,

> Hmm, but we should still avoid marking the DM device as `SYSTEMD_READY=1` in case we have any uevent before DM dev activation

`SYSTEMD_READY=1` is equivalent with `SYSTEMD_READY` not set at all, see above.

>  (so before it has any table loaded and resumed, e.g. the genuine `add` event for the DM device). This is not covered by suggested changes for the `99-systemd.rules`, is it?

No. This is a new aspect that our previous conversation didn't cover.

> For example, `dmsetup create testdev1 --table "0 1 zero`:
> [...]

So you're saying that `SYSTEMD_READY` must be set to 0 for `add` events. My first patch had removed the current code that does this, and would therefore be wrong.
My thinking here had been that a dm device that has just been added would have zero interesting properties for systemd (`ID_FS_TYPE` etc), and thus it wouldn't do any harm if systemd considered it "ready".

If this is wrong, we also can't blindly import `SYSTEMD_READY` from the db. We must be able to figure out that 3) is the event we actually need to react on. Note that 3) may happen before switching root, and that we may see a coldplug event instead after switching root:

```
4)
ACTION=add
DM_UDEV_DISABLE_OTHER_RULES_FLAG=
DM_UDEV_PRIMARY_SOURCE_FLAG=1
DM_ACTIVATION=1
DM_UDEV_DISABLE_OTHER_RULES_FLAG=
```

We have to be able to distinguish the genuine "add" event (type 1) above) from this kind of synthetic "add" events.
Should we do this instead?

```
SUBSYSTEM=="block", ACTION=="add", DM_UDEV_DISABLE_OTHER_RULES_FLAG="1", ENV{SYSTEMD_READY}=="0"
SUBSYSTEM=="block", ACTION=="add", DM_UDEV_DISABLE_OTHER_RULES_FLAG!="1", ENV{SYSTEMD_READY}==""
```

I think we can't, because if a coldplug event was received while the device is suspended, we'd mistakenly set `SYSTEMD_READY="0"`, which might cause the device to be unmounted.

What about this? In the initial "add" event, we won't have a `DM_UUID` set.
```
SUBSYSTEM=="block", ACTION=="add", KERNEL=="dm-*", ENV{DM_UUID}!=".*", ENV{SYSTEMD_READY}=="0"
SUBSYSTEM=="block", DM_UDEV_DISABLE_OTHER_RULES_FLAG="1", IMPORT{db}="SYSTEMD_READY"
```

**Comment by prajnoha** [comment]
> So you're saying that `SYSTEMD_READY` must be set to 0 for `add` events. My first patch had removed the current code that does this, and would therefore be wrong. My thinking here had been that a dm device that has just been added would have zero interesting properties for systemd (`ID_FS_TYPE` etc), and thus it wouldn't do any harm if systemd considered it "ready".

Well, we have to ignore that `add` event (coming before the table load+resume) because the `change` event that follows (for the load+resume) may have the flags decoded from the `DM_COOKIE` (exactly like the example with `lvcreate` that marks the device as "yet unusable" because it needs to do the initialization first - zeroing/wiping). We can't pass through the `DM_COOKIE` for the (genuine) `add` event - that event is out of DM driver's control.

> 
> If this is wrong, we also can't blindly import `SYSTEMD_READY` from the db. We must be able to figure out that 3) is the event we actually need to react on.

We do have the `DM_ACTIVATION` already to cover:
  - the event that notifies about DM device being activated right now (the load+resume scenario)
  - synthetic `add` event that comes anytime after we already have the DM device activated (that's the coldplug at boot scenario)

However, I'm having doubts that it would be enough to check for the `DM_ACTIVATION` only (this is usually for indicating when a layer above should react). Instead, what we would probably need is a variable clearly stating "the DM device is now ready to use" and keep this variable until the end.

Simply, what we need to achieve is a rule set like (DUDORF is DM_UDEV_DISABLE_OTHER_RULES_FLAG):

  A) if a DM device hasn't ever been activated before, then SYSTEMD_READY=0 (covers the case when we have the `add` event before we do any load+resume) (here the DUDORF=1 is always set)

  B) if a DM device has been activated and it still *continually* keeps the DUDORF=1 set, then also keep SYSTEMD_READY=0 (covers the case when a device still needs further initialization even after activation, like the LV scenario)

  C) whenever the DUDORF is dropped after activation (or set to 0), then SYSTEMD_READY=1

  D) whenever the DUDORF=1 is set after activation, then IMPORT{db}=SYSTEMD_READY

It's crappy, yes, but unfortunately, we have only a single `change` event to cover all the substates we need for a DM dev (I suppose the very similar scheme applies for MD too).

> Note that 3) may happen before switching root, and that we may see a coldplug event instead after switching root:
> 
> ```
> 4)
> ACTION=add
> DM_UDEV_DISABLE_OTHER_RULES_FLAG=
> DM_UDEV_PRIMARY_SOURCE_FLAG=1
> DM_ACTIVATION=1
> DM_UDEV_DISABLE_OTHER_RULES_FLAG=
> ```
> 
> We have to be able to distinguish the genuine "add" event (type 1) above) from this kind of synthetic "add" events. 

Synthetic uevents now always have the `SYNTH_UUID` variable set (since kernel version 4.13 I think) - so we can make use of this (the DM udev rules also try to detect synthetic uevents, even without the `SYNTH_UUID` variable as the rules are older - it's detecting whether they have passed the `add` --> `change with DM_COOKIE` sequence).

Also note that there might event be a synthetic uevent generated in between like:
  1) `add` event (DM device creation)
  2) synthetic event (whatever action)
  3) `change` event (DM device table load+resume)

So we need to be cautious, but that example is an extreme, I have to admit. But it's possible.

Should we do this instead?
> 
> ```
> SUBSYSTEM=="block", ACTION=="add", DM_UDEV_DISABLE_OTHER_RULES_FLAG="1", ENV{SYSTEMD_READY}=="0"
> SUBSYSTEM=="block", ACTION=="add", DM_UDEV_DISABLE_OTHER_RULES_FLAG!="1", ENV{SYSTEMD_READY}==""
> ```
> 
> I think we can't, because if a coldplug event was received while the device is suspended, we'd mistakenly set `SYSTEMD_READY="0"`, which might cause the device to be unmounted.
> 
> What about this? In the initial "add" event, we won't have a `DM_UUID` set.
> 
> ```
> SUBSYSTEM=="block", ACTION=="add", KERNEL=="dm-*", ENV{DM_UUID}!=".*", ENV{SYSTEMD_READY}=="0"
> SUBSYSTEM=="block", DM_UDEV_DISABLE_OTHER_RULES_FLAG="1", IMPORT{db}="SYSTEMD_READY"
> ```

We can't use `DM_UUID` - the uuid for DM devices is not mandatory (LVM and mpath have them set, but not DM devs in general).

**Comment by prajnoha** [comment]
> Simply, what we need to achieve is a rule set like (DUDORF is DM_UDEV_DISABLE_OTHER_RULES_FLAG):
> A) if a DM device hasn't ever been activated before, then SYSTEMD_READY=0 (covers the case when we have the `add` event before we do any load+resume) (here the DUDORF=1 is always set)
> B) if a DM device has been activated and it still _continually_ keeps the DUDORF=1 set, then also keep SYSTEMD_READY=0 (covers the case when a device still needs further initialization even after activation, like the LV scenario)
> C) whenever the DUDORF is dropped after activation (or set to 0), then SYSTEMD_READY=1
> D) whenever the DUDORF=1 is set after activation, then IMPORT{db}=SYSTEMD_READY

For that to work, I think we always need to set `SYSTEMD_READY` for a DM device, never keep it "implicit", just to avoid confusion for the rules application.

**Comment by mwilck** [comment]
@prajnoha 

> We can't use DM_UUID - the uuid for DM devices is not mandatory (LVM and mpath have them set, but not DM devs in general).

Ok, but we can use `DM_NAME`. There is no dm device without a name. We might also use `DM_UDEV_PRIMARY_SOURCE_FLAG`, but I suppose that's rather an internal one.

> A) if a DM device hasn't ever been activated before, then SYSTEMD_READY=0 (covers the case when we have the add event before we do any load+resume) (here the DUDORF=1 is always set)

We could cover this with `DM_NAME`.

B) if a DM device has been activated and it still continually keeps the DUDORF=1 set, then also keep SYSTEMD_READY=0 (covers the case when a device still needs further initialization even after activation, like the LV scenario)

This can be handled with `IMPORT{db}`

C) whenever the DUDORF is dropped after activation (or set to 0), then SYSTEMD_READY=1

We can't ask systemd (or any other subsystem for that matter) to remember DUDORF and compare it to it's previous value. If anything, the dm layer would need to do this, and set some property to indicate the "DUDORF has just been cleared" condition. But I would prefer a different solution.

In the dm layer, we use `DM_UDEV_PRIMARY_SOURCE_FLAG=1` to indicate that the device has been properly set up. Other layers could use it, too, but I'm not sure if this flag is considered internal. However, `DM_NAME` is set if and only if `DM_UDEV_PRIMARY_SOURCE_FLAG=1` is, and thus I think we could use it.

D) whenever the DUDORF=1 is set after activation, then IMPORT{db}=SYSTEMD_READY

I see no difference between B) and D).

> For that to work, I think we always need to set SYSTEMD_READY=0/1 for a DM device, never keep it "implicit", just to avoid confusion for the rules application.

I don't like this idea much, mostly because `ENV{SYSTEMD_READY}="1"` falsely suggests to the reader of the code that something is being activated. The only reason to do this is the `IMPORT{db}` weirdness mentioned above.

With the above reasoning, here's my new suggestion that would IMO cover your case A)-D) above:

```
SUBSYSTEM=="block", ACTION=="add", KERNEL=="dm-*", ENV{DM_NAME}!=".*", ENV{SYSTEMD_READY}=="0"
SUBSYSTEM=="block", DM_UDEV_DISABLE_OTHER_RULES_FLAG=="1", IMPORT{db}="SYSTEMD_READY"
```

**Comment by prajnoha** [comment]
> Ok, but we can use `DM_NAME`. There is no dm device without a name.

Right.

> > A) if a DM device hasn't ever been activated before, then SYSTEMD_READY=0 (covers the case when we have the add event before we do any load+resume) (here the DUDORF=1 is always set)
> 
> We could cover this with `DM_NAME`.

Yup.

> > B) if a DM device has been activated and it still continually keeps the DUDORF=1 set, then also keep SYSTEMD_READY=0 (covers the case when a device still needs further initialization even after activation, like the LV scenario)
> 
> This can be handled with `IMPORT{db}`

Yup.

> > C) whenever the DUDORF is dropped after activation (or set to 0), then SYSTEMD_READY=1
> 
...
> In the dm layer, we use `DM_UDEV_PRIMARY_SOURCE_FLAG=1` to indicate that the device has been properly set up. Other layers could use it, too, but I'm not sure if this flag is considered internal. However, `DM_NAME` is set if and only if `DM_UDEV_PRIMARY_SOURCE_FLAG=1` is, and thus I think we could use it.

Yes, good point!

> > D) whenever the DUDORF=1 is set after activation, then IMPORT{db}=SYSTEMD_READY
> 
> I see no difference between B) and D).

Indeed, yes. I was just going through the sequence from the earliest to the latest uevent possibility and I was assigning the SYSTEMD_READY for it, so I didn't notice they're actually same...

> > For that to work, I think we always need to set SYSTEMD_READY=0/1 for a DM device, never keep it "implicit", just to avoid confusion for the rules application.
> 
> I don't like this idea much, mostly because `ENV{SYSTEMD_READY}="1"` falsely suggests to the reader of the code that something is being activated. The only reason to do this is the `IMPORT{db}` weirdness mentioned above.

Well, I don't consider having `ENV{SYSTEMD_READY}="1"` implicit to be much better :) But anyway...

> With the above reasoning, here's my new suggestion that would IMO cover your case A)-D) above:
> 
> ```
> SUBSYSTEM=="block", ACTION=="add", KERNEL=="dm-*", ENV{DM_NAME}!="?*", ENV{SYSTEMD_READY}="0"
> SUBSYSTEM=="block", DM_UDEV_DISABLE_OTHER_RULES_FLAG=="1", IMPORT{db}="SYSTEMD_READY"
> ```

Yup, that looks good.

Though, when I tried that quickly with "lvcreate", which has that "wiping" phase where the first `change` event should be still marked as "not ready" (the "2)" below), I don't see that in the event, I'm not quite sure why yet:

1) 
UDEV  [439.148445] add      /devices/virtual/block/dm-28 (block)
ACTION=add
DM_UDEV_DISABLE_DISK_RULES_FLAG=1
SYSTEMD_READY=0

2)
UDEV  [439.158969] change   /devices/virtual/block/dm-28 (block)
ACTION=change
DM_UDEV_DISABLE_OTHER_RULES_FLAG=1
(but *no* SYSTEMD_READY=0, so implicit SYSTEMD_READY=1 - not OK here)

3)
UDEV  [439.176743] change   /devices/virtual/block/dm-28 (block)
ACTION=change
SYNTH_UUID=0
DM_UDEV_DISABLE_OTHER_RULES_FLAG=
(no SYSTEMD_READY, so implicit SYSTEMD_READY=1 - which is OK here)

**Comment by mwilck** [comment]
> UDEV [439.148445] add /devices/virtual/block/dm-28 (block) ACTION=add DM_UDEV_DISABLE_DISK_RULES_FLAG=1 SYSTEMD_READY=0
> 
> UDEV [439.158969] change /devices/virtual/block/dm-28 (block) ACTION=change DM_UDEV_DISABLE_OTHER_RULES_FLAG=1 (but _no_ SYSTEMD_READY=0, so implicit SYSTEMD_READY=1 - not OK here)

I never see this event at all. What's the trick to make it happen?

**Comment by prajnoha** [comment]
> > UDEV [439.148445] add /devices/virtual/block/dm-28 (block) ACTION=add DM_UDEV_DISABLE_DISK_RULES_FLAG=1 SYSTEMD_READY=0
> > UDEV [439.158969] change /devices/virtual/block/dm-28 (block) ACTION=change DM_UDEV_DISABLE_OTHER_RULES_FLAG=1 (but _no_ SYSTEMD_READY=0, so implicit SYSTEMD_READY=1 - not OK here)
> 
> I never see this event at all. What's the trick to make it happen?

Just plain linear "lvcreate -l1 vg", but it needs wiping enabled (should be enabled by default). Maybe try "lvcreate -l1 vg --wipesignatures y --zero y" to enforce.

**Comment by prajnoha** [comment]
So this is my complete monitor output with changed DM and systemd rules for `lvcreate -l1 vg` (the 2nd uevent with `SEQNUM=3978` should have had SYSTEMD_READY=0, but it's not there for some reason): https://gist.github.com/prajnoha/15fe120f8c578f4a996ef5e74a024151

**Comment by mwilck** [comment]
The reason for your problem was an embarassing typo of mine above. Try this:

```
SUBSYSTEM=="block", ACTION=="add", KERNEL=="dm-*", ENV{DM_NAME}!="?*", ENV{SYSTEMD_READY}="0"
SUBSYSTEM=="block", ENV{DM_UDEV_DISABLE_OTHER_RULES_FLAG}=="1", ENV{SYSTEMD_READY}=="", IMPORT{db}="SYSTEMD_READY"
```

**Comment by prajnoha** [comment]
> The reason for your problem was an embarassing typo of mine above. Try this:
> 
> ```
> SUBSYSTEM=="block", ACTION=="add", KERNEL=="dm-*", ENV{DM_NAME}!="?*", ENV{SYSTEMD_READY}="0"
> SUBSYSTEM=="block", ENV{DM_UDEV_DISABLE_OTHER_RULES_FLAG}=="1", ENV{SYSTEMD_READY}=="", IMPORT{db}="SYSTEMD_READY"
> ```

Yes! I've just noticed exactly the same and fixed that :) It works now!

File: rules.d/99-systemd.rules.in
Comment (poettering) [inline rules.d/99-systemd.rules.in:23]: hmm, what? DM fires multiple "add" events for the same DM device? uh. that's, well, broken?

devices really should show up via "add", then possibly get a series of "change" events (maybe "bind", "unbind" and so on), and eventually a "remove". But a device where the clean codepath results in multiple "add" events is really weird...

Comment (mwilck) [reply rules.d/99-systemd.rules.in:23]: > hmm, what? DM fires multiple "add" events for the same DM device? uh. that's, well, broken?

No, it doesn't. The other "add" events that we must handle are the synthetic/spurious ones, mostly during coldplug.

Comment (poettering) [reply rules.d/99-systemd.rules.in:23]: Our general rule is that whenever appropriate we want that rules that set properties do so in all "positive" verbs the same way, and hence the typical match would be `ACTION!="remove"` rather than `ACTION=="add"`. This rule you are adding contradicts that, please change or explain.

Comment (mwilck) [reply rules.d/99-systemd.rules.in:23]: This is what makes dm events special: we have to act on *kernel change* events and *synthetic add* events. The purpose of the code above is to distinguish the *synthetic add* events (important) from the *kernel add* events (to be ignored).

File: rules.d/99-systemd.rules.in
Comment (poettering) [inline rules.d/99-systemd.rules.in:27]: So, clean rules in our codebase should never have IMPORT{db}, i.e. we want that uevents stand on their own, and come to the same conclusion always. By looking at the db they have a major input from the side, so that properties stick around forever, if they ever were set incorrectly.

If you look at the rules we ship, we never ise IMORT{db} ourselves, with the exception of the DM  support code, this here. If you touch this drop the whole concept please, and determine anew whether the device is ready or not.

File: rules.d/99-systemd.rules.in
Comment (poettering) [inline rules.d/99-systemd.rules.in:23]: this was added in 466266c1728e3b53b7020e4c0165c3e88576b805, @msekletar hopefully knows why.

File: rules.d/99-systemd.rules.in
Comment (poettering) [reply rules.d/99-systemd.rules.in:23]: not sure i follow? so by synthetic you mean the add uecent that userspace triggers via "udevadm trigger --action=add"? i can follow that this will result in multiple "add" events and that's OK i guess. But how does the above rule distinguish these cases, and why?

Comment (poettering) [reply rules.d/99-systemd.rules.in:23]: and why can't the synthetic uevent handling come to the same SYSTEMD_READY conclusions as the original one? why do we need to check the db for that?

**Comment by poettering** [comment]
btw, is there any chance you just take over the relevant rules here? The DM_* udev properties are all DM userspace properties, right? Hence, why not just move them altogether over into LVM sources, and we just drop them? Having some of them in our tree, and some in LVM's tree seems unnecessarily complicated.

The SYSTEMD_READY interface has been pretty stable for many years, LVM could just set this under its own control, but I see no reason why we'd do this here, if LVM can just do that on its own?

File: rules.d/99-systemd.rules.in
Comment (mwilck) [reply rules.d/99-systemd.rules.in:23]: I know. I've studied the history. My patch set keeps this behavior, it just replaces `DM_SUSPENDED` with `DM_UDEV_DISABLE_OTHER_RULES_FLAG`.

Wrt the `GOTO`, that's really a minor thing. As I wrote in the commit message, I think it isn't necessary because the rest of the rules doesn't apply to the devices in question anyway.

File: rules.d/99-systemd.rules.in
Comment (mwilck) [reply rules.d/99-systemd.rules.in:27]: > So, clean rules in our codebase should never have IMPORT{db}, i.e. we want that uevents stand on their own, and come to the same conclusion always. 

That isn't possible for dm. There are situations in which the device can't be probed, while their "ready" status shouldn't change. We can't set `SYSTEMD_READY=0` in this situation, because that might cause umounts or similar destructive actions on systemd's part. The current rule set has been amended multiple times to avoid this kind of issue.

> By looking at the db they have a major input from the side, so that properties stick around forever, if they ever were set incorrectly.

I see your point, but that can't happen here, because the import statement is bound to `DM_UDEV_DISABLE_OTHER_RULES_FLAG`. This flag indicates a temporary condition that will usually be reverted by the next uevent. It is true that one must be carefule with `IMPORT{db}`. If you look at my [dm patch set](https://gitlab.com/lvmteam/lvm2/-/merge_requests/9), you'll see that I have removed a couple of cases where properties were mistakenly imported.

> If you look at the rules we ship, we never ise IMORT{db} ourselves, with the exception of the DM support code, this here. If you touch this drop the whole concept please, and determine anew whether the device is ready or not.

Like I said, we can't. The best we can do, if you insist, is to move the `IMPORT{db}="SYSTEMD_READY"` statement into the dm ruleset. But it would still need to be done in the systemd rule set as well, otherwise the systemd rules would yield wrong results with all but the latest device-mapper rules.

File: rules.d/99-systemd.rules.in
Comment (mwilck) [reply rules.d/99-systemd.rules.in:23]: > But how does the above rule distinguish these cases, and why?

**How**: using `DM_NAME`, which is not set for the initial "add" event.

**Why**: when the kernel "add" event arrives, the device is an empty container with no table, not useful for anything. As @prajnoha wrote [above](https://github.com/systemd/systemd/pull/31661#issuecomment-1983426438), this event must be ignored.

If you want to know the details, see [10-dm.rules](https://gitlab.com/lvmteam/lvm2/-/blob/main/udev/10-dm.rules.in?ref_type=heads).

> and why can't the synthetic uevent handling come to the same SYSTEMD_READY conclusions as the original one? why do we need to check the db for that?

The two events will come to *different* conclusions.

Between the kernel "add" event and the synthetic / coldplug "add" event, there are one ore more "change" events. These can happen for various reasons. The case that matters is a "change" event caused by a dm table (re)load (more precisely, an DM device resume after changing the active table). In this event, the device properties are communicated by the kernel via `DM_COOKIE`. They need to be remembered in the udev db, because the kernel will send the cookie only once. If this change event occurs during initramfs processing, we need a coldplug "add" event to make the device known after pivot root, and we need to remember the properties for that as well (that's one reason why dm devices need `db_persist`).

Further complication arises because the device may be inaccesible when the "reload" event (or the corresponding later coldplug event) is received. In that case, `blkid` or other stacked subsystems can't probe it. You can see in @prajnoha's example above that this is the default case for newly created LVs. The dm properties must be remembered, and the device activated on the first uevent where the dm devices are properly set up *and* the device is accessible.

@poettering, here is a question for you: would any bad happen if we did **not** set `SYSTEMD_READY=0` in such a case? The device wouldn't have any of the properties that are set by `blkid`. You created the commits that first set `SYSTEMD_READY=0` for these situations yourself back in 2011 (b7cf1b68a79833411851fa547ac9b4906207c224, 444d9432358c7ef8d1a85d330089b0027e740612). I suppose you had reasons to do so, but you were working with a dm ruleset that was different from what we have now.

If we find out that setting `SYSTEMD_READY=0` isn't necessary for incompletely initialized dm devices, we can eliminate this code from `99-systemd.rules` completely.

Comment (mwilck) [reply rules.d/99-systemd.rules.in:23]: > This rule you are adding contradicts that, please change or explain.

The rule really only applies to the initial "add" event. I tried to explain that in the comments above the rule.

Comment (mwilck) [reply rules.d/99-systemd.rules.in:23]: I've tried hard to break a system by *not* setting `ENV{SYSTEMD_READY}="0"` in the rule above, and failed. I believe it may actually be possible to skip this rule.

But I think that needs more testing and discussion. It would be a relatively risky change to make. I'd prefer to have this PR merged as-is for now, and to discuss removal of the entire dm-specific code in `99-systemd.rules` separately, perhaps in a follow-up PR, which I would happily submit.

**Comment by mwilck** [comment]
@yuwata, @poettering, AFAICT I have addressed your comments and questions. If there's more, please let me know.

I don't understand the CI autopkgtest failure for "noble", is it caused by my patches, and if yes, what can I do to fix it?

I saw that previous patches to `99-systemd.rules` also changed `test/fuzz/fuzz-udev-rules/systemd.rules`, should I do that as well?

File: rules.d/99-systemd.rules.in
Comment (msekletar) [reply rules.d/99-systemd.rules.in:23]: > this was added in [466266c](https://github.com/systemd/systemd/commit/466266c1728e3b53b7020e4c0165c3e88576b805), @msekletar hopefully knows why.

This was added in order to fix https://bugzilla.redhat.com/show_bug.cgi?id=1985288.

Comment (mwilck) [reply rules.d/99-systemd.rules.in:23]: @msekletar, am I assuming correctly that the fix for that bug was the `IMPORT{db}="SYSTEMD_READY"` for a suspended device? 

My patch keeps this logic, it just replaces the condition on `DM_SUSPENDED` by a `DM_UDEV_DISABLE_OTHER_RULES_FLAG`. `DM_SUSPENDED` implies `DM_UDEV_DISABLE_OTHER_RULES_FLAG`. Would you concur that my change wouldn't re-introduce breakage for that bug?

Comment (mwilck) [reply rules.d/99-systemd.rules.in:23]: To verify, I just tested the reproducer from https://bugzilla.redhat.com/show_bug.cgi?id=1985288 with my patches from this PR applied. The issue from the bug did not occur.

Comment (msekletar) [reply rules.d/99-systemd.rules.in:23]: > To verify, I just tested the reproducer from https://bugzilla.redhat.com/show_bug.cgi?id=1985288 with my patches from this PR applied. The issue from the bug did not occur.

Thank you!

**Review by yuwata — APPROVED** [review]
LGTM.

Though, as already suggested by @poettering, the related rules should be moved to LVM. But, let's do that later in another PR.

**Comment by mwilck** [comment]
I have re-pushed this PR, squashed all commits into one and completely rewritten the commit message. The code is identical.

I originally thought that this might be easier to review in small chunks, but as the chunks are closely related with each other, the opposite is probably true. I hope this makes sense.

Please read the new commit message; I tried to give a concise rationale for the changes in this PR, and the handling of SYSTEMD_READY for device mapper in general, which hopefully addresses @poettering's concerns.

**Review by yuwata — APPROVED** [review]
If this is ready to go, this should be merged before v256-rc1, and I do not want to merge such change on a late RC stage.
I set the rc-blocker tag.
@poettering Could you take the final look?

**Review by msekletar — COMMENTED** [review]
LGTM

**Comment by mwilck** [comment]
FTR, I'll be out of office next week, so if anything urgently needs further clarification or discussion, please notify me soon.

**Review by bluca — APPROVED** [review]

**Comment by bluca** [comment]
Merging as discussed in the team meeting yesterday

**Comment by mwilck** [comment]
Thanks a lot!

**Pull request merged**
